integration/tests/common/block_state.go

```diff
@@ -134,7 +114,7 @@ func (bs *BlockState) WaitForNextFinalized(t *testing.T) *messages.BlockProposal
 func (bs *BlockState) WaitUntilNextHeightFinalized(t *testing.T) *messages.BlockProposal {
 	currentProposed := bs.highestProposed
 	require.Eventually(t, func() bool {
-		return bs.highestFinalized > currentProposed
+		return bs.highestFinalized > currentProposed && bs.finalizedByHeight[currentProposed+1] != nil
```

Comment:
Why do we need the extra condition here?

Reply:
I actually refactored a bit, so now it works on `bs.finalizedByHeight[currentProposed+1] != nil` instead of previous condition `bs.highestFinalized > currentProposed`, and here is the reason:

With recent changes, we relax the conditions on block state tracker accounting for the intrinsic limitations on missing a block especially at earlier heights, hence, `highestFinalized = h` only accounts that block height `h` is both _available_ and _withing confirmed range_, hence has been finalized. However, it does not say anything regarding the availability of any block at heights below `h`. 

Let's consider an example that `currentProposed = 6` and we wait for block `7`. However, in the meantime, the block state tracker receives and finalizes block height `10` without yet receiving `7`. Hence, there could be `nil` dereferencing situations if we try `bs.finalizedByHeight[7]` when `bs.highestFinalized = 10 > 6`. This new refactoring solves that.